build: optionally skip external warnings

vasild opened the pull request:
Add an option to `./configure` to suppress compilation warnings from
external headers. The option is off by default (no change in behavior,
show warnings from external headers).

This option is useful if e.g. Boost or Qt is installed outside of
`/usr/include` (warnings from headers in `/usr/include` are already
suppressed by default) and those warnings stand in the way of compiling
Bitcoin Core with `-Werror[=...]` or they just clutter the build output
too much and make our own warnings hard to spot.

`-isystem /usr/include` bricks GCC's `#include_next`, so we use
`-idirafter` instead. This way we don't have to treat `/usr/include`
specially.

vasild committed:
build: optionally skip external warnings

Add an option to `./configure` to suppress compilation warnings from
external headers. The option is off by default (no change in behavior,
show warnings from external headers).

This option is useful if e.g. Boost or Qt is installed outside of
`/usr/include` (warnings from headers in `/usr/include` are already
suppressed by default) and those warnings stand in the way of compiling
Bitcoin Core with `-Werror[=...]` or they just clutter the build output
too much and make our own warnings hard to spot.

vasild commented:
cc @Empact @hebasto @theuni

vasild commented:
To test that this works as intended plant some warning in a Qt header (if you don't have already!). For example:
```cpp
inline bool f(int iii, unsigned uuu) { return iii < uuu; }
```
in `/qt/installed/here/include/qt5/QtCore/qglobal.h` and see that the warning is gone with `./configure --enable-suppress-external-warnings`.

vasild commented:
We currently compile our code with some warnings explicitly silenced: `-Wno-unused-local-typedef -Wno-deprecated-register -Wno-implicit-fallthrough`.

Currently Bitcoin Core is free of such warnings, but they can sneak in at anytime completely unnoticed - due to the above flags nobody will notice if a new code introduces such a warning.

If this PR gets merged, then we can consider removing the above `-Wno-...` flags. If it turns out that some external header produces such warnings, then they can be silenced using the functionality from this PR. This way we will spot if a new code introduces such a warning in Bitcoin Core.

jonatack commented:
Concept ACK

brakmic commented:
Concept ACK.

hebasto commented:
Tested on Ubuntu 20.04 LTS (gcc 9.3.0) and macOS 10.15.4 (llvm clang 10.0). Works as expected.

I've looked through our docs, and didn't find any compiler warning policy. Or did I miss it?

From the POV of a user who respects one's own safety and builds one's own node binaries from the source, I believe there is no a such thing like a "harmless compiler warning". At least for major OSes (the two latest versions of Ubuntu LTS, Debian stable, Fedora, macOS, Windows with default system compilers) on the x86_64 platform. In the worst case, a rare warning should be documented in the release notes.

Concept ACK.

Is it possible to increase the granularity of `--enable-suppress-external-warnings`?
E.g., `--enable-suppress-external-warnings=qt`?
Related: #16722.

vasild commented:
> Tested on Ubuntu 20.04 LTS (gcc 9.3.0) and macOS 10.15.4 (llvm clang 10.0). Works as expected.

Excellent!
 
> From the POV of a user who respects one's own safety and builds one's own node binaries from the source, I believe there is no a such thing like a "harmless compiler warning".

I agree. This is why the new option is `off` by default. And I do not think it should ever be changed (to `on` by default).

> Is it possible to increase the granularity of `--enable-suppress-external-warnings`?
> E.g., `--enable-suppress-external-warnings=qt`?

I did consider this! It would increase the complexity of this feature to some extent, but there is a bigger problem - we suppress warnings based on the directory in which the header which produced the warning is located. So, if both qt and boost are installed in the same directory (e.g. `/usr/local/include`) then such separation will not work as expected.

vasild commented:
Rebased (trivial conflict with the merged #18297).

vasild commented:
Rebased to resolve conflicts.

vasild commented:
Rebased to resolve conflicts.

practicalswift commented:
ACK 272066393dd3a70df7fc889cea95e1c7c943fc1a

Thanks for doing this @vasild: having this in `master` would allow us to enable additional compiler diagnostics without having to worry about warning spam from external dependencies :)

hebasto commented on configure.ac:
```diff
@@ -1170,6 +1188,14 @@ else
 
   dnl sets $bitcoin_enable_qt, $bitcoin_enable_qt_test, $bitcoin_enable_qt_dbus
   BITCOIN_QT_CONFIGURE([5.5.1])
+
+  dnl Keep a copy of the original $QT_INCLUDES and use it when invoking qt's moc
+  QT_INCLUDES_UNSUPPRESSED=$QT_INCLUDES
+  if test x$suppress_external_warnings != xno ; then
+    QT_INCLUDES=SUPPRESS_WARNINGS($QT_INCLUDES)
+    QT_DBUS_INCLUDES=SUPPRESS_WARNINGS($QT_DBUS_INCLUDES)
+    QT_TEST_INCLUDES=SUPPRESS_WARNINGS($QT_TEST_INCLUDES)
+  fi
```
Should these lines be a part of `Makefile.qt.include`?

vasild replied:
I don't think it is possible to call functions defined in `configure.ac` from a `Makefile`.

hebasto commented on configure.ac:
```diff
@@ -1149,6 +1159,14 @@ AC_SUBST(LEVELDB_CPPFLAGS)
+dnl See -idirafter documentation:
+dnl https://gcc.gnu.org/onlinedocs/gcc/Directory-Options.html
+dnl https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-idirafter-arg
+AC_DEFUN([SUPPRESS_WARNINGS], [`echo $1 |sed -E -e 's/(^| )-I/\1-idirafter /g'`])
```
nit: [Use $(...) notation instead of legacy backticked \`...\`.](https://github.com/koalaman/shellcheck/wiki/SC2006)

vasild replied:
Done.

hebasto reviewed (commented):
Tested 272066393dd3a70df7fc889cea95e1c7c943fc1a on top of the master (b99a1633b270e0e89479b2bb2ae19a8a8dc0fa05) on Linux Mint 20 (x86_64, focal code base).

```
$ make -C depends CC=clang CXX='clang++ -stdlib=libc++' NO_QT=1
$ CONFIG_SITE=$PWD/depends/x86_64-pc-linux-gnu/share/config.site ./configure --enable-suppress-external-warnings CC=clang CXX='clang++ -stdlib=libc++'
$ make clean
$ make
$ lldb src/bitcoind -- -regtest
...
Process 366548 stopped
* thread #1, name = 'bitcoind', stop reason = signal SIGSEGV: invalid address (fault address: 0x0)
    frame #0: 0x00007ffff7cb14e5 libc.so.6`__strlen_avx2 at strlen-avx2.S:65
(lldb) bt
* thread #1, name = 'bitcoind', stop reason = signal SIGSEGV: invalid address (fault address: 0x0)
  * frame #0: 0x00007ffff7cb14e5 libc.so.6`__strlen_avx2 at strlen-avx2.S:65
    frame #1: 0x0000555555b51453 bitcoind`__os_strdup + 35
    frame #2: 0x0000555555b2b847 bitcoind`__env_add_data_dir + 199
    frame #3: 0x0000555555b2b722 bitcoind`__env_set_data_dir + 18
    frame #4: 0x0000555555afbe99 bitcoind`DbEnv::set_data_dir(char const*) + 57
    frame #5: 0x0000555555957676 bitcoind`BerkeleyEnvironment::Open(bilingual_str&) + 950
    frame #6: 0x0000555555958f12 bitcoind`BerkeleyDatabase::Verify(bilingual_str&) + 210
    frame #7: 0x000055555595d63c bitcoind`MakeBerkeleyDatabase(boost::filesystem::path const&, DatabaseOptions const&, DatabaseStatus&, bilingual_str&) + 524
    frame #8: 0x000055555594354b bitcoind`MakeDatabase(boost::filesystem::path const&, DatabaseOptions const&, DatabaseStatus&, bilingual_str&) + 619
    frame #9: 0x00005555558fb624 bitcoind`MakeWalletDatabase(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DatabaseOptions const&, DatabaseStatus&, bilingual_str&) + 180
    frame #10: 0x000055555585001c bitcoind`VerifyWallets(interfaces::Chain&) + 2892
    frame #11: 0x0000555555839e3b bitcoind`interfaces::(anonymous namespace)::WalletClientImpl::verify() + 27
    frame #12: 0x00005555555b7139 bitcoind`AppInitMain(util::Ref const&, NodeContext&, interfaces::BlockAndHeaderTipInfo*) + 5177
    frame #13: 0x00005555555a2776 bitcoind`main + 2550
    frame #14: 0x00007ffff7b4d0b3 libc.so.6`__libc_start_main(main=(bitcoind`main), argc=2, argv=0x00007fffffffdd78, init=<unavailable>, fini=<unavailable>, rtld_fini=<unavailable>, stack_end=0x00007fffffffdd68) at libc-start.c:308:16
    frame #15: 0x00005555555a1cbe bitcoind`_start + 46
```

vasild commented:
Updated: use `-isystem` instead of `-idirafter` because if there are two incompatible boost installations in the system, for example in `/usr/include/boost` and in `/whatever/boost`, then `-I/whatever` or `-isystem /whatever` will cause the second one to be picked (correct) while `-idirafter /whatever` will cause the first one to be picked from `/usr/include` (incorrect). Thanks, @hebasto!

Also added BDB to the list.

hebasto reviewed (commented):
Approach ACK 359870798edb77474a57536e5b13512b2d388253, tested on macOS 10.15.7 with Apple clang 11.0.3 and Qt 5.15.1, and Linux Mint 20 (focal codebase) with gcc and clang.

It would be nice to list some system setups when `--enable-suppress-external-warnings` _does_ suppress warning, as all of the systems I tested emit only warnings that I've [injected](https://github.com/bitcoin/bitcoin/pull/18750#issuecomment-618894244) myself :)

... or list [planned](https://github.com/bitcoin/bitcoin/pull/18750#issuecomment-691622922) additional compiler diagnostics.

vasild commented:
> It would be nice to list some system setups ...

You mean here on GitHub? Here is one:
FreeBSD 12, all defaults (boost 1.72, Qt 5.15.0, bdb 5.3.28), without this option 27519 warnings are printed. Some of them:

```
/usr/local/include/boost/system/error_code.hpp:290:18: warning: 'name' overrides a member function but is not marked 'override' [-Wsuggest-override]

/usr/local/include/db5/db_cxx.h:1403:22: warning: 'what' overrides a member function but is not marked 'override' [-Wsuggest-override]

/usr/local/include/qt5/QtWidgets/qstackedwidget.h:54:5: warning: 'qt_metacall' overrides a member function but is not marked 'override' [-Wsuggest-override]
```

vasild commented:
Another example:
https://travis-ci.org/github/bitcoin/bitcoin/jobs/730973642#L2748

```
/home/travis/build/bitcoin/bitcoin/depends/arm-linux-gnueabihf/share/../include/boost/process/detail/posix/executor.hpp:191:16: warning: ignoring return value of ‘ssize_t write(int, const void*, size_t)’, declared with attribute warn_unused_result [-Wunused-result]

         ::write(_pipe_sink, &len, sizeof(int));

         ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

This patch would allow us to use `--enable-werror` on `ARM [GOAL: install] [buster] [unit tests, functional tests]` in CI so that newly added warnings on this platform in Bitcoin Core are detected (cause CI failure).

practicalswift commented:
ACK 359870798edb77474a57536e5b13512b2d388253: patch looks correct

luke-jr commented on configure.ac:
```diff
@@ -1149,6 +1159,18 @@ AC_SUBST(LEVELDB_CPPFLAGS)
+dnl is not necessary (/usr/include is already a system directory) and because
+dnl it would break GCC's #include_next.
+AC_DEFUN([SUPPRESS_WARNINGS],
+         [$(echo $1 |sed -E -e 's/(^| )-I/\1-isystem /g' -e 's|-isystem /usr/include|-I/usr/include|g')])
```
Should use `${SED}` here in place of `sed`

vasild replied:
Done

luke-jr commented on configure.ac:
```diff
@@ -1149,6 +1159,18 @@ AC_SUBST(LEVELDB_CPPFLAGS)
+dnl is not necessary (/usr/include is already a system directory) and because
+dnl it would break GCC's #include_next.
+AC_DEFUN([SUPPRESS_WARNINGS],
+         [$(echo $1 |sed -E -e 's/(^| )-I/\1-isystem /g' -e 's|-isystem /usr/include|-I/usr/include|g')])
```
This incorrectly excludes `/usr/includefoo`

Perhaps use `s#-isystem /usr/include([/ ]|$)#-I/usr/include\1#g`

vasild replied:
Done, except that the `#` separator causes the following during `autogen.sh`:

```
configure.ac:11: error: m4_init: unbalanced m4_divert_push:
configure.ac:1199: m4_divert_push: GROW
configure.ac:11: m4_divert: BODY
configure.ac:1171: SUPPRESS_WARNINGS is expanded from...
configure.ac:1171: SUPPRESS_WARNINGS is expanded from...
configure.ac:1171: SUPPRESS_WARNINGS is expanded from...
configure.ac:1171: SUPPRESS_WARNINGS is expanded from...
configure.ac:1171: SUPPRESS_WARNINGS is expanded from...
configure.ac:11: the top level
autom4te-2.69: /usr/local/bin/gm4 failed with exit status: 1
aclocal: error: echo failed with exit status: 1
autoreconf-2.69: aclocal failed with exit status: 1
```

so I used `;` as a separator instead.

hebasto reviewed (approved):
ACK ba8950ee0134a7958e3e9b041cd54d222feb09a1, tested on Linux Mint 20 (x86_64).

With `--enable-suppress-external-warnings` having warnings from leveldb only:
```
$ make > /dev/null 
leveldb/db/c.cc: In function ‘leveldb_filterpolicy_t* leveldb_filterpolicy_create_bloom(int)’:
leveldb/db/c.cc:474:17: warning: ‘virtual const char* leveldb_filterpolicy_create_bloom(int)::Wrapper::Name() const’ can be marked override [-Wsuggest-override]
  474 |     const char* Name() const { return rep_->Name(); }
      |                 ^~~~
leveldb/db/c.cc:475:10: warning: ‘virtual void leveldb_filterpolicy_create_bloom(int)::Wrapper::CreateFilter(const leveldb::Slice*, int, std::string*) const’ can be marked override [-Wsuggest-override]
  475 |     void CreateFilter(const Slice* keys, int n, std::string* dst) const {
      |          ^~~~~~~~~~~~
leveldb/db/c.cc:478:10: warning: ‘virtual bool leveldb_filterpolicy_create_bloom(int)::Wrapper::KeyMayMatch(const leveldb::Slice&, const leveldb::Slice&) const’ can be marked override [-Wsuggest-override]
  478 |     bool KeyMayMatch(const Slice& key, const Slice& filter) const {
      |          ^~~~~~~~~~~
```

luke-jr reviewed (approved):
utACK ba8950ee0134a7958e3e9b041cd54d222feb09a1

practicalswift commented:
ACK ba8950ee0134a7958e3e9b041cd54d222feb09a1: diff looks correct!

Really looking forward to having this in `master` :)

laanwj commented:
Thanks for adding this option, it definitely helps reduce the flood of warnings in boost and qt etc.
